src/main/java/com/mparticle/kits/KochavaKit.java

```diff
@@ -110,37 +129,60 @@ public void removeUserIdentity(MParticle.IdentityType identityType) {
         return null;
     }
 
-    private void createKochava() {
-        if (feature == null) {
-            HashMap<String, Object> datamap = new HashMap<String, Object>();
-            datamap.put(Feature.INPUTITEMS.KOCHAVA_APP_ID, getSettings().get(APP_ID));
-            datamap.put(Feature.INPUTITEMS.APP_LIMIT_TRACKING, getSettings().get(LIMIT_ADD_TRACKING));
-            datamap.put(Feature.INPUTITEMS.DEBUG_ON, Boolean.parseBoolean(getSettings().get(ENABLE_LOGGING)));
-            datamap.put(Feature.INPUTITEMS.REQUEST_ATTRIBUTION, Boolean.parseBoolean(getSettings().get(RETRIEVE_ATT_DATA)));
-            if (identityLink != null) {
-                datamap.put(Feature.INPUTITEMS.IDENTITY_LINK, identityLink);
-            }
-            feature = new Feature(getContext(), datamap);
-
-        }
-    }
-
     @Override
     public List<ReportingMessage> setOptOut(boolean optOutStatus) {
-        if (feature != null) {
-            feature.setAppLimitTracking(optOutStatus);
-            List<ReportingMessage> messageList = new LinkedList<ReportingMessage>();
-            messageList.add(
-                    new ReportingMessage(this, ReportingMessage.MessageType.OPT_OUT, System.currentTimeMillis(), null)
-                            .setOptOut(optOutStatus)
-            );
-            return messageList;
-        }else {
-            return null;
-        }
+        Tracker.setAppLimitAdTracking(optOutStatus);
+        List<ReportingMessage> messageList = new LinkedList<>();
+        messageList.add(
+                new ReportingMessage(this, ReportingMessage.MessageType.OPT_OUT, System.currentTimeMillis(), null)
+                        .setOptOut(optOutStatus)
+        );
+        return messageList;
     }
 
     public static void setIdentityLink(Map<String, String> identityLink) {
         KochavaKit.identityLink = identityLink;
     }
+
+    private void setAttributionResultParameter(String key, JSONObject value) {
+        AttributionResult result = getKitManager().getAttributionResults().get(getConfiguration().getKitId());
```

Comment:
I don't think other kits do this so not sure if it's needed here - I would just create a new attribution result every time otherwise it could get confusing when customers get callbacks with a mix of results.

Reply:
gotcha, I was thinking the tradeoff would be that the data you get from the polling API method, `MParticle.getAttributionResults()` would be incomplete, but I suppose the listener is a much better interface for them to be using anyways